### DISCUSSION TITLE
[Feature Request] Warm instance pool

### DISCUSSION THREAD

**Discussion by sds** [runs-on/runs-on#171] [Ideas]
First off: incredible work here. The simplicity with getting started and having builds just work brought a tear to my eye.

I see from the [KISS documentation](https://runs-on.com/about/#keep-it-stupid-simple) avoiding warm pools etc. has been intentional, which makes sense.

Would you be open to a pull request adding support? Goal would be:

- Opt-in feature
- Global in scope (i.e. you define how many "warm" instances of each type that you have in the pool—they're not allocated to any specific repo)
- Instances would be hibernated after boot to reduce cost, and started by runs-on once an appropriate job was ready
- Let AWS ASG logic handle the rest (maintaining the pool as instances are removed + terminated, etc.)

Trying to get a sense of how strongly we want to hang on to the KISS principle for this project. Thanks for your consideration!

**Comment by crohr** [comment]
Hi @sds, and thanks for the nice words!

My main objection for now about pools is that this would force a very specific kind of runner (architecture, image, disk size, spot vs non-spot, etc.) because nothing can be changed at runtime. But I can see how this could be useful anyway as a base runner for short-lived jobs etc. Is that your use-case as well?

Just a few bullet-points/questions that come to mind:
- as you said, available at the organization level. Requires permissions update to the app.
- allow multiple pools or just one? Defined within the CF stack (declarative), or from `.github-private/runs-on.yml`? Probably CF stack so that we get to see if the changes are successful, and less stuff to put in the RunsOn server.
- fixed size pool, with configurable number of standby (hibernated) instances? Or do you also need auto-scaling? I wonder what is the marginal cost for each additional standby instance, since you only pay for storage cost, and with hibernated we wouldn't need to pay for provisioned throughput since the AMI would already be loaded.
- how to handle rollout of ASG when base image is updated. Probably need some lambda somewhere that updates the ASG settings.

So all in all I'm really interested in at least having a stab at it. Next step would be to dive a bit deeper into how it would all work once a workflow_job queued event comes in, assuming we have an ASG of say 10 hibernated instances.

**Comment by sds** [comment]
Thanks for your response! 

Agree, I want to make sure we have a plan that aligns before writing any code. Quick clarifying question, and then I'll respond in detail to your other questions:

>because nothing can be changed at runtime

I might be misunderstanding. With ASGs, you can list arbitrary attributes to specify the kind of instances you want in the ASG—see [Created mixed instances group using attribute-based instance type selection](https://docs.aws.amazon.com/autoscaling/ec2/userguide/create-mixed-instances-group-attribute-based-instance-type-selection.html)). This gives you a fair amount of flexibility for a particular "weight class" of instances (e.g. # cores, or minimum memory) if you aren't particular about the actual instance type.

That said, I agree that it does require the operator to give a bit of thought as to the kinds of instances they need for their jobs.

Let me respond to your questions and get your feedback on a draft proposal.

---

Re: multiple pools—this would be my desired outcome, but let's scope down to a single pool for now for a proof of concept, if that works for you. Most devs would likely still appreciate a single pool of small instances that boot a little faster since they were already pre-booted.

Re: fixed size pool—Fixed capacity for now, yes. Would set `DesiredCapacity` at 0 and `MaxGroupPreparedCapacity` to whatever operators want so that no instances are running and the prepared set are hibernated. A default of 1 seems fine, but there's no harm in increasing since you'll eventually end up using all prepared instances anyway.

Re: base image updating—seems reasonable to respond to an event when the AMI changes. Suggestion of triggering a lambda seems reasonable (I haven't looked closely at the CloudFormation stack yet).

Let me know if any of the above requires clarification.

---

Since there's at least a little interest, give me some time to look at the code and get more familiar. I'll follow up with a description of how a job would flow through the system in this new world.

If anyone (including yourself, Cyril) wants to take a stab, don't wait up on me! I might not get to this this week, but I'm excited to dig in!

**Comment by crohr** [comment]
> I might be misunderstanding. With ASGs, you can list arbitrary attributes to specify the kind of instances you want in the ASG—see [Created mixed instances group using attribute-based instance type selection](https://docs.aws.amazon.com/autoscaling/ec2/userguide/create-mixed-instances-group-attribute-based-instance-type-selection.html)). This gives you a fair amount of flexibility for a particular "weight class" of instances (e.g. # cores, or minimum memory) if you aren't particular about the actual instance type.
> 
> That said, I agree that it does require the operator to give a bit of thought as to the kinds of instances they need for their jobs.

What I meant is that you can't change the instance definition based on your job labels, once the stack has been created/updated, compared to the current behaviour. But yes that constraint is probably fine for pools.

**Comment by sds** [comment]
Hey @crohr, I had the embarrassing realization that the server component is closed source (understandably—your sponsorship-level license explains). 

So with that said, I'm not able to validate how difficult this proposal would be with respect to the necessary changes in the server. The proposal is different than what I initially described above, but makes it much easier to use without worrying about unnecessary expense or knowing your runner requirements _a priori_. Appreciate you bearing with me as I make some assumptions—thanks in advance for any feedback!

---

_**Note**: I tried to follow [this format](https://github.com/runs-on/runs-on/pull/164/files#diff-26d40f63ae7916a74eae6fbaf7ee56dea525f5d0e3db352cc5a7008a2d5ed1b6)_

# Automatic creation of warm instance Pools

Initial EC2 instance startup time can be avoided if instances are booted before they are needed, since (especially with encrypted volumes) the pages are loaded into cache, and other AWS resources are already allocated (ENIs, etc.) rather than requested on-demand.

We propose a solution where for any given runner, a user can specify whether they want to preserve a warm pool of instances as jobs are enqueued with that runner. Hibernation will be enabled by default to save on cost (reducing it to just EBS volume storage for the duration of the hibernated instance's existence), but there is an optional setting to change the behavior to actually stop warm pool instances instead of hibernating, though this increases startup time and reduces the overall benefit.

_**Note**: AWS supports hibernation for a subset of instance families and regions, falling back to stopping the instance if not supported—see [docs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/hibernating-prerequisites.html#hibernation-prereqs-supported-amis)._

## How it works

If a warm pool is enabled for a given runner, at the time the server requests an EC2 instance it does the following:
- Checks if a warm pool ASG is already created
- If not, creates it using the configuration specified (see below)
- If already created, detaches an instance from the ASG and starts it.

The ASG will automatically take care of provisioning a new warm instance to replace the detached instance.

This allows warm pools to be lazily created as the user defines them in their GitHub workflows, which has the nice benefit of not requiring the user to have to identify which runner configurations they need up front.

### Avoiding ASGs

While not required/suggested for a first draft, given the logic we desire is relatively simple, the use of ASGs at all could be avoided. The server component can identify all hibernated EC2 instances that are candidates for a given pool, and if it finds one allocate it and immediately issue a replacement based on configuration. That said, we may want to stick with ASGs since they provide a convenient audit trail in the AWS console for when instances were provisioned and consumed.

## How to configure

When specifying runner constraints:
- if `warm-pool-size=N` is specified, after the first instance is requested create an ASG with `MaxGroupPreparedCapacity=N`, `MinGroupPreparedCapacity=N` and `DesiredSize=0`.
- if `warm-pool-transition={stop,hibernate}` is specified, behavior of the ASG's warm pool is changed (default is `hibernate`)
- if `warm-pool-name=...` is specified, the ASG will be shared among all jobs that request the same warm pool name. If the other runner constraints don't match, an error is thrown in the workflow

When reusing warm pools across jobs, it is recommended that you define the runner constraints as a GitHub Actions Variable (repo scope if you are going to use across multiple jobs in the same repo, or organization scope if you are going to use in jobs across multiple repositories in an organization).

![Example reusable pool configuration](https://github.com/user-attachments/assets/fba7c28c-05ae-4de6-9f1c-2c11d2b2dcea)

You can then reference this pool in jobs via:

```yaml
runs-on: runs-on=${{ github.run_id }},$RUNNER_WARM_POOL_2CPU_LINUX_ARM64
```

This allows you to define the constraints in a single place and update them across all your jobs without needing to merge (potentially many) PRs.

To avoid "leaking" ASGs, a daily cleanup utility will be added that inspects the last instance creation time of all warm pool ASGs, and if the last instance was created over a week ago (configurable via CloudFormation), treats the configuration as defunct and deletes it. While this isn't perfect, it ensures that as runner configurations change, ASGs for combinations that aren't used within the specified interval are eventually cleaned up, helping limit unnecessary spend.

**Comment by crohr** [comment]
@sds ah yes indeed the server component is private for now. I thought you were part of a company that bought the sponsorship license and could see the code :)

I'm a bit wary about creating ASGs on-demand (quota issues + as you said, ensuring its lifecycle / cleanup), or maintaining fleets on hibernated instances manually. Also not sure how it would work if you have multiple branches, and alter the runner configuration for the same pool name in one of them.

I think I would still ase a solution on ASGs, because at least it can easily handle the rollout of instances with updated AMIs, as well as the scale-up / scale-down of resources. I'm thinking we could also use a CloudFormation stack for each warm pool, so that the definition and maintenance of those would be separate from the main stack (i.e. opt-in feature, better observability, and possibility to have N pools). And then RunsOn would just regularly refresh the pool names available from the list of pool stacks.

I'm moving this issue to the discussions panel for now, in case anyone has more input for us!

Also happy to give you access to the server code if you want!

**Comment by sds** [comment]
Wanted to add this observation: one warm pool might suffice for multiple different instance types, since the [instance type can be changed](https://depot.dev/blog/faster-ec2-boot-time#resizing-the-instance) before the instance is started again.

Of course there are limitations (instance store vs regular EBS volume instances), but the flexibility does mean a single ASG could pre-provision all instances assuming the user isn't using a specialized set of instances.

**Comment by Nuru** [comment]
I suggest you look at the the [Pool](https://github.com/github-aws-runners/terraform-aws-github-runner/blob/4750ae183fd75e0c21c3fd5a48ce5ce7446edc3b/docs/configuration.md#pool) and [Idle runners](https://github.com/github-aws-runners/terraform-aws-github-runner/blob/4750ae183fd75e0c21c3fd5a48ce5ce7446edc3b/docs/configuration.md#idle-runners) sections of the Philips Labs [GitHub Runners on AWS](https://github-aws-runners.github.io/terraform-aws-github-runner/) documentation for inspiration and guidance on how to design this feature.

RunsOn is unique in being able to configure runners in the Workflow via `runs-on` labels. All other implementations require pre-specified runner configurations, along the lines of what RunsOn provides as "Default runners". Idle pools would probably have to be designated by such labels, although they could pick up any job for which they match the job's requirements. 

In my limited experience, `2cpu-linux-x64` runners are sufficient for most jobs that would be tasked to a pool for which idle runners were wanted. In practice, companies have jobs they want sped up and jobs they don't, and they can easily arrange for job runner labels and idle pool runners to match up.

**Comment by crohr** [comment]
Another issue with pools is that those probably make sense at an org level, while RunsOn currently registers runners at the repository level. This allows better isolation across repositories but I don't think having pools for each repo would make sense economically. Unless most usage would happen in a single repo anyway?

**Reply by Nuru** [reply]
Org runners would be useful, and should be on the roadmap, but idle pools for repo runners would still be better than nothing. It just means you would have to make the configuration of the pool specific to a repo. 

I can imagine customers that would want idle pools even if limited to 1 repo, such as their main app source code repo.

**Comment by crohr** [comment]
The feature has now landed in v2.9+: https://runs-on.com/runners/pools/.
